Polish the CorsFilter based on resteasy + add it to cli

RemiDesgrange opened the pull request.

danduk82 reviewed (approved):
lgtm, untested

RemiDesgrange commented:
Tests are failing I need to investigate since it was working on my machine.

bchapuis reviewed (commented):
Looks good to me, but we should probably add resteasy’s header licence (ASL) to the source file. What do you thinking?

RemiDesgrange commented:
I have no idea because I didn't copy paste the entire code and rewrote some part. Maybe I should add a comment and a link to the resteasy file saying that I was inspired by their implemementation?

bchapuis commented:
I faced a similar situation with MIT code elsewhere in the code base and retained the copyright even if the code evolved. https://github.com/baremaps/baremaps/blob/main/baremaps-postgres/src/main/java/com/baremaps/postgres/jdbc/CopyWriter.java

Notice that we shouldn't integrate portions of GPL code as it is more restrictive than ASL, MIT or BSD.

bchapuis commented:
Regarding the tests, it is failing with the integration tests are you running mvn install -P integration?

RemiDesgrange commented:
I figured it out just after writing my comments. I'll work on it a bit to fix the test.

RemiDesgrange commented:
* I added the reateasy licence header.
* I removed the test for CORS because jdk11 restrict setting some http header (https://github.com/AdoptOpenJDK/openjdk-jdk8u/blob/master/jdk/src/share/classes/sun/net/www/protocol/http/HttpURLConnection.java#L276-286) It saddens me in every possible way...

bchapuis commented:
Interesting, I havn't used the new HTTP client very often. Do you think this is because CORS headers are a browser thing?

RemiDesgrange commented:
I have no idea why these headers are restricted... Do you want me to use another HTTP client lib ? I'm pretty rusty when it comes to choosing Java libs.

bchapuis commented:
The openapi module uses Jersey Test. Here is an example produced by @b4l:
https://github.com/baremaps/baremaps/blob/tilesets/baremaps-openapi/src/test/java/com/baremaps/openapi/services/StylesServiceTest.java

In general, when writing basic tests, I simply use HTTPUrlConnection which allows to set custom headers:

```
URL url = new URL("http://example.com/");
HttpURLConnection conn = (HttpURLConnection) url.openConnection();
```

bchapuis commented on baremaps-server/src/main/java/com/baremaps/server/CorsFilter.java:
```diff
@@ -1,28 +1,63 @@
+    if (origin == null || requestContext.getMethod().equalsIgnoreCase("OPTIONS") || requestContext.getProperty("cors.failure") != null) {
+      return;
+    }
+    responseContext.getHeaders().putSingle(CorsHeaders.ACCESS_CONTROL_ALLOW_ORIGIN, origin);
```
We may move CorsHeaders' constants in the CorsFilter class to make these lines shorter (java's verbosity...), but it is also fine for me as it is.

```java
var headers = responseContext.getHeaders();
headers.putSingle(ACCESS_CONTROL_ALLOW_ORIGIN, origin);
headers.putSingle(ACCESS_CONTROL_ALLOW_METHODS, "GET, POST, PUT, DELETE, OPTIONS, HEAD");
...

RemiDesgrange replied:
I didn't know I could do that !

RemiDesgrange commented on baremaps-cli/src/test/java/com/baremaps/cli/BaremapsTest.java:
```diff
@@ -97,14 +97,12 @@ void test() throws InterruptedException, IOException {
 
     // Wait for ServiceTalk and JAX-RS to respond with 200 OK
-    await().timeout(60, TimeUnit.SECONDS).pollDelay(5, TimeUnit.SECONDS).until(() -> {
+    await().timeout(30, TimeUnit.SECONDS).pollDelay(5, TimeUnit.SECONDS).until(() -> {
```
ouuups sorry...

RemiDesgrange commented:
> In general, when writing basic tests, I simply use HTTPUrlConnection which allows to set custom headers:
> 
> ```
> URL url = new URL("http://example.com/");
> HttpURLConnection conn = (HttpURLConnection) url.openConnection();
> ```

As you can I'm using the "new" Http client that appear in JDK9: https://github.com/baremaps/baremaps/pull/310/files#diff-79032f8befc30a3655833a26b20cb85c4851129484c47c0cf06651f329583316R125 

it's quite elegant and works well
